### PR TITLE
feat: Support multiple env for one user

### LINKED ISSUE #18: feat(api): Refactor environments API
## Description

Use /users/<>/environments post to create the environment.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by gaocegege**
Close #18 

This PR is to:

- Support multiple envs per user

Signed-off-by: Ce Gao <cegao@tensorchord.ai>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -45,7 +45,7 @@ REGISTRY ?= ghcr.io/tensorchord
 BASE_REGISTRY ?= docker.io
 
 # Disable CGO by default.
-CGO_ENABLED ?= 1
+CGO_ENABLED ?= 0
```
Comment (VoVAllen) [inline Makefile:48]: Why is this changed?

File: sshname/name.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+package sshname
+
+import (
+	"fmt"
+	"strings"
+)
+
+func Username(owner, envName string) (string, error) {
+	return fmt.Sprintf("%s/%s", owner, envName), nil
```
Comment (VoVAllen) [inline sshname/name.go:13]: Does this mean user will use `allen/mnist` as the ssh username?

File: Makefile (hunk 1 above)
Comment (gaocegege) [reply Makefile:48]: It is to run on ARM.

File: pkg/server/containerssh.go [hunk 3]
```diff
@@ -61,16 +69,23 @@ func (s *Server) OnPubKey(c *gin.Context) {
...
 		key, _, _, _, err := ssh.ParseAuthorizedKey([]byte(req.PublicKey.PublicKey))
 		if err != nil {
 			logrus.WithError(err).Error("failed to parse key")
 			c.JSON(500, err)
 			return
 		}
 		// https://github.com/golang/go/issues/21704#issuecomment-342760478
-		if bytes.Equal(key.Marshal(), k.PublicKey.Marshal()) {
+		if owner == k.IdentityToken &&
+			bytes.Equal(key.Marshal(), k.PublicKey.Marshal()) {
```
Comment (kemingy) [inline pkg/server/containerssh.go:88]: I think it should use https://pkg.go.dev/crypto/subtle#ConstantTimeCompare

Comment (gaocegege) [reply pkg/server/containerssh.go:88]: SGTM

**Comment by gaocegege** [comment]
PTAL @VoVAllen @kemingy

File: pkg/server/containerssh.go [hunk 4]
```diff
@@ -61,16 +69,23 @@ func (s *Server) OnPubKey(c *gin.Context) {
...
 		key, _, _, _, err := ssh.ParseAuthorizedKey([]byte(req.PublicKey.PublicKey))
 		if err != nil {
 			logrus.WithError(err).Error("failed to parse key")
 			c.JSON(500, err)
 			return
 		}
 		// https://github.com/golang/go/issues/21704#issuecomment-342760478
-		if bytes.Equal(key.Marshal(), k.PublicKey.Marshal()) {
+		if owner == k.IdentityToken &&
+			subtle.ConstantTimeCompare(key.Marshal(), k.PublicKey.Marshal()) == 0 {
```
Comment (kemingy) [inline pkg/server/containerssh.go:88]:
```suggestion
			subtle.ConstantTimeCompare(key.Marshal(), k.PublicKey.Marshal()) == 1 {
```
> ConstantTimeCompare returns 1 if the two slices, x and y, have equal contents

Comment (gaocegege) [reply pkg/server/containerssh.go:88]: Great catch!

**Review by VoVAllen — APPROVED** [review]
LGTM

**Comment by VoVAllen** [comment]
One potential feature in the future: users can specify hardware requirements or pod templates through the username. For example, user can use `allen/mnist/t4` vs `all/mnist/v100` to support running environments on different underlying hardware.

**Comment by kemingy** [comment]
> One potential feature in the future: users can specify hardware requirements or pod templates through the username. For example, user can use `allen/mnist/t4` vs `all/mnist/v100` to support running environments on different underlying hardware.

Could be a feature of the dashboard.

**Review by kemingy — APPROVED** [review]
/lgtm
